control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go

```diff
@@ -413,32 +413,43 @@ func getNodesToUpgrade(nodes []*corev1.Node, targetConfig string, maxUnavailable
 	// - NodeNetworkUnavailable condition status is ConditionFalse.
 	// TODO (jerzhang): consider what we want to do with node status here
 	// For now, we will just check current/desired config to see if any nodes is already updating
+
 	var numUnavailable int
+	var candidateNodes []*corev1.Node
 	for _, node := range nodes {
 		if node.Annotations[CurrentMachineConfigAnnotationKey] != node.Annotations[DesiredMachineConfigAnnotationKey] {
 			numUnavailable++
+			// Compared to self-driving OCP, there is an additional scenario to consider here.
+			// Since the ConfigMap contents are synced separately, those will change on the fly in the pod.
+			// Meaning that we could have a scenario where there are multiple queue'ed updates, in which case
+			// the MCD will just jump straight to the latest version.
+			// This will cause the MCD to softlock, so let's make sure for those unavailable nodes, we are also
+			// update their desired configuration. The MCD should be able to reconcile these changes on the fly.
+			if node.Annotations[DesiredMachineConfigAnnotationKey] != targetConfig {
+				candidateNodes = append(candidateNodes, node)
+			}
 		}
 	}
 
 	capacity := maxUnavailable - numUnavailable
-	// If we're at capacity, there's nothing to do.
+	// If we're at capacity, just return the list of already updating nodes populated by the above loop
 	if capacity < 1 {
-		return nil
+		return candidateNodes
 	}
-	// We only look at nodes which aren't already targeting our desired config
-	var candidateNodes []*corev1.Node
+	// We only look at nodes which aren't already targeting our desired config,
+	// and do not have an ongoing update
 	for _, node := range nodes {
-		if node.Annotations[DesiredMachineConfigAnnotationKey] != targetConfig {
+		if node.Annotations[DesiredMachineConfigAnnotationKey] != targetConfig &&
+			node.Annotations[CurrentMachineConfigAnnotationKey] == node.Annotations[DesiredMachineConfigAnnotationKey] {
 			candidateNodes = append(candidateNodes, node)
+			capacity--
```

Comment:
Is this change vs `return candidateNodes[:capacity]` just cosmetic?

Reply:
Not exactly. With the new changes, I would essentially be returning `candidateNodes[:capacity]` appended with `nodes that already unavailable (don't count towards capacity) but don't have the latest desired config`.

I could have instead created two slices to track those, and concatenated them at the end here, but I thought that was a bit unecessary. So now candidateNodes tracks all the nodes that can be updated, and we just directly return it, instead of looping through all the nodes, most of which won't be selected anyways